### PR TITLE
add grpc_metadata_scrubber filter

### PR DESCRIPTION
**Pull request by qiwzhang**
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

This PR is ready for review.

It only has c++ code changes to add the filter.  Next PR will have configmgr changes to add this filter

### CODE REVIEW

File: api/envoy/v8/http/grpc_scrubber/config.proto [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package espv2.api.envoy.v8.http.grpc_scrubber;
```
Comment (JLXIA) [inline api/envoy/v8/http/grpc_scrubber/config.proto:17]: -> grpc_metadata_scrubber?

Comment (qiwzhang) [reply api/envoy/v8/http/grpc_scrubber/config.proto]: Done

File: src/envoy/http/grpc_metadata_scrubber/filter_test.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  testing::NiceMock<MockStreamEncoderFilterCallbacks> mock_cb_;
+};
+
+TEST_F(GrpcMetadataScrubberFilterTest, ContentLengthRemoved) {
+  Envoy::Http::TestResponseHeaderMapImpl headers{{"content-type", "application/grpc"},
+                                                {"content-length", "100"}};
+  EXPECT_EQ(Envoy::Http::FilterHeadersStatus::Continue,
+            filter_->encodeHeaders(headers, false));
+
+  EXPECT_EQ(1L, Envoy::TestUtility::findCounter(mock_factory_context_.scope_,
```
Comment (nareddyt) [inline src/envoy/http/grpc_metadata_scrubber/filter_test.cc:54]: Can we flip around the assertions for consistency? `EXPECT_EQ(x, 1L)`

go/c-readability-advice#avoid-yoda-style-comparisons

**Review by nareddyt — APPROVED** [review]

File: src/envoy/http/grpc_metadata_scrubber/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+## Overview
+
+This filter checks response headers; if content-type is "application/grpc", if the response
+header has content-length, remove content-length header.
+
+This is needed in some special deployment cases with followings:
+* downstream uses http1 to transport gRPC requests and uses trunk encoding to support response trailers.
+* upstream adds content-length
+* envoy drops the response trailers if the response headers have content-length.
```
Comment (TAOXUY) [inline src/envoy/http/grpc_metadata_scrubber/README.md:11]: Can you specify when envoy drops the response trailers? 
Is there any filter sequence requirement?

File: src/envoy/http/grpc_metadata_scrubber/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# gRPC Metadata Scrubber Filter
+
+## Overview
+
+This filter checks response headers; if content-type is "application/grpc", if the response
+header has content-length, remove content-length header.
+
+This is needed in some special deployment cases with followings:
+* downstream uses http1 to transport gRPC requests and uses trunk encoding to support response trailers.
+* upstream adds content-length
```
Comment (TAOXUY) [inline src/envoy/http/grpc_metadata_scrubber/README.md:10]: period

**Review by TAOXUY — APPROVED** [review]

File: src/envoy/http/grpc_metadata_scrubber/README.md (hunk 4 above)
Comment (qiwzhang) [reply src/envoy/http/grpc_metadata_scrubber/README.md]: Done

File: src/envoy/http/grpc_metadata_scrubber/README.md (hunk 3 above)
Comment (qiwzhang) [reply src/envoy/http/grpc_metadata_scrubber/README.md]: Done

File: src/envoy/http/grpc_metadata_scrubber/filter_test.cc (hunk 2 above)
Comment (qiwzhang) [reply src/envoy/http/grpc_metadata_scrubber/filter_test.cc]: Done

**Review by nareddyt — APPROVED** [review]

**Pull request merged**